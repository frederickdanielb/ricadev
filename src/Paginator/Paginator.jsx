import { useEffect, useState } from "react";

const Paginator = ({ total, limit, activePage, onChangePage }) => {
  
    const [lastPage, setlastPage] = useState(0);

    const getLastPage = (items) => items / limit;

    useEffect(() => {
        setlastPage(getLastPage(total));
        onChangePage(0);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [total]);

    const selectPage = (evt) => {
        const selectedItem = Number.parseInt(evt.target.dataset.item);
        onChangePage(selectedItem);
    };
    const prevPage = () => {
        const selectedItem = activePage -1;
        if (selectedItem >= 0) onChangePage(selectedItem);
    };
    const nextPage = () => {
        const selectedItem = activePage +1;
        if (selectedItem <= lastPage) onChangePage(selectedItem);
    };

    const PageItems = () => {
        let PageItems = [];
        for (let index = (activePage-2<0?0:activePage-2); index < ((activePage-2+limit)>lastPage?lastPage:(activePage-2+limit)); index++) {
            PageItems.push(
                <li className="page-item" key={"page-item-li"+index}>
                    <div
                        key={"page-item-"+index}
                        className={
                            "page-item " +
                            (index === activePage ? "active" : "")
                        }
                    >
                        <button
                            className="page-link"
                            onClick={selectPage}
                            data-item={index}
                        >
                            {index + 1}
                        </button>
                    </div>
                </li>
            );
        }
        return PageItems;
    };
    return (
        <nav
            aria-label="Page navigation example"
            className="pagination-section mt-0"
        >
            <ul className="pagination">
                <li className="page-item">
                    <button
                        onClick={prevPage}
                        className="page-link"
                        aria-label="Previous"
                    >
                        <span aria-hidden="true">&laquo;</span>
                        <span className="sr-only">Anterior</span>
                    </button>
                </li>
                {PageItems()}
                <li className="page-item">
                    <button
                        onClick={nextPage}
                        className="page-link"
                        aria-label="Next"
                    >
                        <span aria-hidden="true">&raquo;</span>
                        <span className="sr-only">Siguiente</span>
                    </button>
                </li>
            </ul>
        </nav>
    );
};

export default Paginator;
