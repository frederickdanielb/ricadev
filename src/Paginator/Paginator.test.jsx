import { fireEvent, render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import Paginator from "./Paginator";

describe("Paginator component", () => {
    const onChange = vi.fn();
	it("should render correctly", async () => {
		const { asFragment } = render(<Paginator onChangePage={onChange}  activePage={0} limit={10} total={100}  />);

		/* act(() => {
			screen.getByText("weather");
		});
		await waitFor(() =>
			expect(screen.getAllByText(/9.7/)[0]).toBeInTheDocument()
		); */
		expect(asFragment()).toMatchSnapshot();

        act(() => {
            const nextButton = screen.getByText("Siguiente");
            fireEvent.click(nextButton);
        })
        act(() => {
            const lastButton = screen.getByText("Anterior");
            fireEvent.click(lastButton);
        })
        act(() => {
            const numberButton = screen.getByText("3");
            fireEvent.click(numberButton);
        })
        expect(onChange).toHaveBeenCalledTimes(3);
	});
});
