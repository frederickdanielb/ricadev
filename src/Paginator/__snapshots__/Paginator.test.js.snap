// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component should render correctly 1`] = `
<DocumentFragment>
  <nav
    aria-label="Page navigation example"
    class="pagination-section mt-0"
  >
    <ul
      class="pagination"
    >
      <li
        class="page-item"
      >
        <button
          aria-label="Previous"
          class="page-link"
        >
          <span
            aria-hidden="true"
          >
            «
          </span>
          <span
            class="sr-only"
          >
            Anterior
          </span>
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item active"
      >
        <button
          class="page-link"
          data-item="0"
        >
          1
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="1"
        >
          2
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="2"
        >
          3
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="3"
        >
          4
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="4"
        >
          5
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="5"
        >
          6
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="6"
        >
          7
        </button>
      </li>
      <li
        class="page-item"
      />
      <li
        class="page-item "
      >
        <button
          class="page-link"
          data-item="7"
        >
          8
        </button>
      </li>
      <li
        class="page-item"
      >
        <button
          aria-label="Next"
          class="page-link"
        >
          <span
            aria-hidden="true"
          >
            »
          </span>
          <span
            class="sr-only"
          >
            Siguiente
          </span>
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
